### ISSUE TITLE
Nemotron 3 SFT example uses target_modules that may not exist in hybrid architecture

### ISSUE THREAD

**Issue by davidnichols-ops** [huggingface/trl#6773]
## Problem

TRL's official Nemotron 3 SFT example (`examples/scripts/sft_nemotron_3.py`) uses LoRA target modules that may not exist in NemotronH's hybrid architecture:

```python
target_modules=["q_proj", "k_proj", "v_proj", "o_proj", 
                "gate_proj", "up_proj", "down_proj"]
```

The `gate_proj`, `up_proj`, `down_proj` modules are standard MLP projection names from Llama-style architectures. NemotronH uses Mamba-2 layers (with `in_proj`, `out_proj`) and MoE layers (with different naming). These MLP module names may silently not match anything in the model.

## Impact

Users following the official example may get LoRA adapters that only target a small fraction of the model's layers (the attention projections), with the MLP/Mamba modules silently skipped. This produces the same "loss not improving" failure mode we experienced:

- DPO LoRA training: 25 steps, loss stuck at 0.6931 (ln(2))
- Zero behavioral change in the model
- No warning that target modules didn't match

## Reproduction

We ran DPO LoRA on `nvidia/NVIDIA-Nemotron-Nano-9B-v2` with attention-only target modules (same as the example minus the MLP modules). The LoRA adapter only touched 4 out of 56 layers (7.1%). Loss never moved from ln(2). Full writeup: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

## Proposed Fix

1. Update the example to use target modules that actually exist in NemotronH:

```python
# Inspect the model to find valid target modules
import re
target_modules = list(set(
    name.split(".")[-1] for name, module in model.named_modules()
    if isinstance(module, torch.nn.Linear) and 
    re.match(r".*\d+\..*", name)  # only layer-level modules
))
```

Or explicitly list NemotronH-appropriate modules:

```python
target_modules=[
    # Attention
    "q_proj", "k_proj", "v_proj", "o_proj",
    # Mamba (if PEFT allows — see PEFT issue about mamba_model_types ban)
    "in_proj", "out_proj", 
    # MoE/MLP — check actual module names in the model
    # "gate", "up", "down"  # verify these exist before including
]
```

2. Add a comment in the example explaining that NemotronH is a hybrid architecture and users should verify target modules with `model.named_modules()`.

3. Consider adding a runtime check in TRL that warns when configured `target_modules` don't match any modules in the model.

## Environment

- trl: 1.10.0
- transformers: 5.16.0.dev0
- Model: nvidia/NVIDIA-Nemotron-Nano-9B-v2

## Related

- PEFT PR #3289: Added default target_modules for nemotron_h (attention only)
- PEFT issue (filed separately): NemotronH LoRA target modules silent failure
- Full writeup: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

**Comment by davidnichols-ops** [comment]
<p dir="ltr"><strong>Cheatsheet: LoRA target modules for NemotronH (hybrid Mamba-Attention)</strong></p>
<p dir="ltr">Leaving this here since a few people (including me) have hit silent LoRA failures on this model family. Summary of what's confirmed:</p>
<div dir="ltr">
Module | Exists in NemotronH? | Safe LoRA target? | Notes
-- | -- | -- | --
q_proj, k_proj, v_proj, o_proj | Yes (attention layers) | Yes | Standard attention projections
gate_proj | No | N/A | NemotronH MLP has no gating (up_proj/down_proj only) — PEFT silently no-ops on the name mismatch, no error
up_proj, down_proj | Yes (MLP layers) | Yes |  
in_proj | Yes (Mamba layers) | Yes | Called as a module in forward(), works fine with LoRA
out_proj | Yes (Mamba layers) | No — banned by PEFT | _check_lora_target_modules_mamba in tuners_utils.py blocks it; the fused SSM kernel consumes out_proj.weight as a raw tensor, bypassing the LoRA wrapper. Silent no-op in training, shape-mismatch crash at merge_and_unload()

</div>
<p dir="ltr"><strong>Corrected target list:</strong></p>
<div role="group" aria-label="Code" tabindex="0"><div><div></div></div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre-wrap;">q_proj k_proj v_proj o_proj in_proj up_proj down_proj</code></pre></div></div>
<p dir="ltr"><strong>Why this matters:</strong> in Nemotron-Nano-9B-v2, 52 of 56 layers are Mamba blocks. Any target list that omits <code>in_proj</code> leaves ~93% of the model frozen — LoRA "trains" but nothing happens (matches the ln(2)-stuck-loss symptom above). Worth checking <code>model.named_modules()</code> against your target list on any new hybrid architecture rather than assuming Llama-style names carry over.</p>
<p dir="ltr"><strong>Open question for maintainers:</strong> would a runtime warning when a configured <code>target_modules</code> entry matches zero modules be welcome in TRL/PEFT generally? Seems like it'd catch this whole class of bug (not just NemotronH) cheaply. </p>

**Comment by behroozazarkhalili** [comment]
I checked this against the model the example names, `nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16`. One of the three modules is dead, the other two are not.

Counting module suffixes in the checkpoint's `model.safetensors.index.json` (6243 tensors, 6151 distinct modules), which is the same final-path-component rule PEFT uses to match a list-valued `target_modules`:

| target | modules matched |
| --- | --- |
| `q_proj`, `k_proj`, `v_proj`, `o_proj` | 6 each |
| `up_proj`, `down_proj` | 2967 each |
| `gate_proj` | **0** |

So `up_proj` and `down_proj` do match, and they match a lot: `config.json` has `n_routed_experts=128` and `n_shared_experts=1`, and the `hybrid_override_pattern` has 23 `M` (Mamba-2), 23 `E` (MoE) and 6 `*` (attention) across the 52 layers. That gives 23 x (128 + 1) = 2967 expert MLP projections and 6 of each attention projection, which is exactly what the index shows. Instantiating `NemotronHForCausalLM` from Transformers and listing its `nn.Linear` leaves gives the same set: `in_proj`, `out_proj`, `q_proj`, `k_proj`, `v_proj`, `o_proj`, `up_proj`, `down_proj`, `lm_head`.

`gate_proj` is the one entry that matches nothing, and it is silent because PEFT raises only when *no* entry matches. On PEFT 0.18.1, against a module tree with `up_proj` and `down_proj` but no `gate_proj`, the list `["q_proj", "up_proj", "down_proj", "gate_proj"]` adapts the first three and reports nothing about the fourth, while `["gate_proj"]` alone raises `ValueError`.

Opened #6866 to drop just that name. It does not change which modules get adapted, since PEFT was already ignoring it; it removes a target that does not exist for this model from the example and from the adapter config a run would save.

One thing the numbers raise that the fix does not address: the 23 Mamba-2 mixers (`in_proj`, `out_proj`) get no adapter at all, so 23 of 52 layers have an unadapted token-mixing path. Whether the example should target them is a training-design question rather than a naming bug, and I have no measurement either way, so I left it alone.

**Comment by behroozazarkhalili** [comment]
Correction to my earlier count. On Transformers 5.11.0 the 128 routed experts of each MoE layer are one fused `NemotronHExperts` module with stacked `weight` parameters, not per-expert `nn.Linear` layers, so PEFT's `target_modules` cannot reach them. `up_proj` and `down_proj` therefore match the 23 shared-expert projections, not 2967 modules; the 2967 I reported counted checkpoint tensors, not modules. The conclusion is unchanged: both names are live targets and `gate_proj` is the only dead one. #6866 now carries the corrected numbers and also removes `gate_proj` from the notebook cell.
